### PR TITLE
Cbc UI d2 628 adding keysets

### PR DESCRIPTION
**Pull request by cody-constine-ttd**

### CODE REVIEW

File: src/main/java/com/uid2/shared/store/IKeysetProvider.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.uid2.shared.store;
+
+import java.time.Instant;
+
+public interface IKeysetProvider {
```
Comment (jon8787) [inline src/main/java/com/uid2/shared/store/IKeysetProvider.java:5]: Do we actually need this interface? RotatingKeysetProvider implements it but IKeysetProvider doesn't seem to be used anywhere else

File: src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public List<KeysetKey> getActiveKeysetKeys() {
+        return this.activeKeys;
+    }
+
+    @Override
+    public KeysetKey getActiveKey(int keysetId, Instant now) {
+        List<KeysetKey> keysetKeys = keysetMap.get(keysetId);
+        if(keysetKeys == null || keysetKeys.isEmpty()) return null;
+        int it = Utils.upperBound(keysetKeys, now, (ts, k) -> ts.isBefore(k.getActivates()));
```
Comment (aulme) [inline src/main/java/com/uid2/shared/store/KeysetKeyStoreSnapshot.java:33]: Why is it called `it`? Can there be a more useful name?

File: src/main/java/com/uid2/shared/store/IKeysetProvider.java (hunk 1 above)
Comment (cody-constine-ttd) [reply src/main/java/com/uid2/shared/store/IKeysetProvider.java:5]: I was just copying how ACLs work. Looks like operator uses it but doesn't implement it, so I think we are probably good to remove it and just use the RotatingKeysetProvider

File: src/main/java/com/uid2/shared/store/reader/RotatingKeysetProvider.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Map;
+
+public class RotatingKeysetProvider implements StoreReader<Map<Integer, Keyset>> {
+    private final ScopedStoreReader<KeysetSnapshot> reader;
+
+    public RotatingKeysetProvider(DownloadCloudStorage fileStreamProvider, StoreScope scope) {
+        this.reader = new ScopedStoreReader<>(fileStreamProvider, scope, new KeysetParser(), "keysets");
+    }
+
+    public KeysetSnapshot getSnapshot(Instant asOf) {
```
Comment (Ian-Nara) [inline src/main/java/com/uid2/shared/store/reader/RotatingKeysetProvider.java:22]: What is the point of overriding with the asOf paramter?

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        keysets.put("location", location);
+        metadata.put("keysets", keysets);
+        return metadata;
+    }
+
+    @Test
+    public void loadsContentToSiteScope() throws Exception{
+        RotatingKeysetProvider provider = new RotatingKeysetProvider(cloudStorage, new SiteScope(new CloudPath("metadata"), 5));
+        JsonArray content = new JsonArray();
+        addKeyset(content, 5, 1, 1, 2, 3);
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:75]: these addKeysets are hard to read because of the list next to the other params. Maybe List.Of() would make them clearer?

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    @Mock private ICloudStorage cloudStorage;
+    private RotatingKeysetProvider keysetProvider;
+
+    @Before
+    public void setup() {
+        mocks = MockitoAnnotations.openMocks(this);
+        keysetProvider = new RotatingKeysetProvider(cloudStorage, new GlobalScope(new CloudPath("metadata")));
+    }
+
+    private void addKeyset(JsonArray content, int siteId, int keysetId, int... listedSiteIds) {
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:45]: should listedSiteIds be called allowedSites?

File: src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        return new KeysetKey(0, null, null, null, null, keysetId);
+    }
+    private boolean canAccessKey(int clientSiteId, int keySiteId) {
+        return keysetProvider.getSnapshot().canClientAccessKey(makeClientKey(clientSiteId), makeKey(keySiteId));
+    }
+
+    @Test
+    public void loadContentMultipleEntries() throws Exception {
+        JsonArray content = new JsonArray();
+        addKeyset(content, 1, 1, 2,3);
```
Comment (jon8787) [inline src/test/java/com/uid2/shared/store/reader/RotatingKeysetProviderTest.java:128]: - should we add tests where allowed sites is empty?
- a test where an allowed sites list has duplicate site ids?
